apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/internal/ApolloCacheInterceptor.kt

```diff
@@ -134,26 +137,38 @@ internal class ApolloCacheInterceptor(
       /**
        * This doesn't use [readFromNetwork] so that we can publish all keys all at once after the keys have been rolled back
        */
-      var response: ApolloResponse<D>? = null
-      var exception: ApolloException? = null
-      try {
-        response = chain.proceed(request).single()
-      } catch (e: ApolloException) {
-        exception = e
-      }
+      var networkException: ApolloException? = null
+      val networkResponse: Flow<ApolloResponse<D>> = chain.proceed(request)
+          .catch {
+            if (it is ApolloException) {
+              networkException = it
+            } else {
+              throw it
+            }
+          }
 
-      val optimisticKeys = if (optimisticData != null) {
-        store.rollbackOptimisticUpdates(request.requestUuid, publish = false)
-      } else {
-        emptySet()
-      }
+      var optimisticKeys: Set<String>? = null
+
+      networkResponse.collect { response ->
+        if (optimisticKeys == null) optimisticKeys = if (optimisticData != null) {
+          store.rollbackOptimisticUpdates(request.requestUuid, publish = false)
+        } else {
+          emptySet()
+        }
 
-      if (response != null) {
-        maybeWriteToCache(request, response, customScalarAdapters, optimisticKeys)
+        maybeWriteToCache(request, response, customScalarAdapters, optimisticKeys!!)
         emit(response)
-      } else {
-        store.publish(optimisticKeys)
-        throw exception!!
+      }
+
+      if (networkException != null) {
+        if (optimisticKeys == null) optimisticKeys = if (optimisticData != null) {
```

Comment:
I'm sorry I have to repeat this block (also not 100% certain this is correct). Could there be a nicer way to write this?

Reply:
The more I think about it, the more I think optimistic updates + defer doesn't make sense:

```graphql
mutation {
  updateUser(name: $name, email: $email) {
    name
    email @defer
  }
}
```

What if we set `"John Doe"`/`"john@doe.com"` as optimistic data and we get the first payload but the second fails? We're now in a weird state where the local cache doesn't match the backend because we can't rollback individual fields. 

For a first version, I would throw at runtime if there is optimistic data but downstream emits several responses. We can always revisit and introduce "incremental" optimistic data (but I hope we won't have to 😄 )

Reply:
Hmm I see! Indeed that is not ideal and we can forbid this case for now 👍